calico_node/tests/st/config/test_felix_config.py

```diff
@@ -0,0 +1,410 @@
+            if passed:
+                _log.info("check access success!%s", msg)
+                return
+            else:
+                _log.exception("retry [%s] check access failed!%s", retry, msg)
+                time.sleep(1)
+        self.fail(("check access failed!%s", msg))
+
+    @staticmethod
+    def check_access(expect_access, host_src, host_target, port, protocal="tcp"):
+        protocol_opt = "-u" if protocal == "udp" else ""
+        cmd = "echo \"hostname && exit\" | timeout -t 3 nc %s -w 1 %s %s" % \
+              (protocol_opt, host_target.ip, port)
+        remote_hostname = host_src.execute(cmd, raise_exception_on_failure=False)
+        hostname = host_target.execute("hostname")
+        _log.info("check access from %s to %s:%s %s, result is |%s|, "
+                  "target hostname |%s|, expect access %s",
+                  host_src.ip, host_target.ip, port, protocal,
+                  remote_hostname, hostname, expect_access)
+
+        # If port is 179, we are talking to bird.
+        # If bird returns with a valid string, this means connection is successful.
+        if port == 179 and remote_hostname is not None and len(remote_hostname) > 0:
+            remote_hostname = hostname
+
+        result = (hostname == remote_hostname)
+        return result == expect_access
+
+    def add_policy(self, policy_data):
+        self._apply_resources(policy_data, self.host1)
+
+    def remove_host_endpoint(self):
+        self.delete_all("hostEndpoint")
+
+    @classmethod
+    def delete_all(cls, resource):
+        # Grab all objects of a resource type
+        objects = yaml.load(cls.hosts[0].calicoctl("get %s -o yaml" % resource))
+        # and delete them (if there are any)
+        if len(objects) > 0:
+            _log.info("objects: %s", objects)
+            if 'items' in objects and len(objects['items']) == 0:
+                pass
+            else:
+                cls._delete_data(objects, cls.hosts[0])
+
+    @classmethod
+    def _delete_data(cls, data, host):
+        _log.debug("Deleting data with calicoctl: %s", data)
+        cls._exec_calicoctl("delete", data, host)
+
+    @classmethod
+    def _apply_resources(cls, resources, host):
+        cls._exec_calicoctl("apply", resources, host)
+
+    @staticmethod
+    def _exec_calicoctl(action, data, host):
+        # Delete creationTimestamp fields from the data that we're going to
+        # write.
+        for obj in data.get('items', []):
+            if 'creationTimestamp' in obj['metadata']:
+                del obj['metadata']['creationTimestamp']
+        if 'metadata' in data and 'creationTimestamp' in data['metadata']:
+            del data['metadata']['creationTimestamp']
+
+        # Use calicoctl with the modified data.
+        host.writefile("new_data",
+                       yaml.dump(data, default_flow_style=False))
+        host.calicoctl("%s -f new_data" % action)
+
+    @classmethod
+    def get_container_ip(cls, container_name):
```

Comment:
Does 'container' here mean Workload or docker Host?  Do we already have this function somewhere?  Maybe ip is already a property of the workload or host?

Reply:
Function not used in this test. Removed.